update recommended build instructions (#908)
LLNL/serac: src/docs/sphinx/quickstart.rst

Building Serac
--------------

Serac is a CMake project (tested with v3.18). Configure with at least the ``-C``

> flag below, to ensure CMake can locate the dependencies installed by spack:

Reviewer: ```suggestion
flag below, to ensure CMake can locate the dependencies installed by Spack:
```

.. code-block:: bash

  $ cmake -B build \
          -C path/to/serac/<machine>-<type>-<compiler>.cmake \
          -DCMAKE_BUILD_TYPE=Release \
          -S path/to/serac

LLNL developers working on one of the LC machines can instead use one of 

> the provided CMake cache files in serac's ``host-configs`` directory:

Reviewer: ```suggestion
the provided CMake initial cache files, also referred to as a host-config, in Serac's ``host-configs`` directory:
```

``-C path/to/serac/host-configs/<machine>-<type>-<compiler>.cmake``

----------
 

> Some other frequently-useful configure flags

Reviewer: ```suggestion
Some other frequently useful configure flags:
```

``-B <build-directory>`` (required) 

>   tells CMake what to name the build directory

Reviewer: ```suggestion
  Tells CMake what to name the build directory
```

``CMAKE_BUILD_TYPE`` (recommended) one of ``Debug`` (default), ``Release``, ``RelWithDebInfo``, ``MinSizeRel``

>   these options describe whether or not to compile with debug info or optimization flags

Reviewer: ```suggestion
  These options describe whether or not to compile with debug info or optimization flags
```

``SERAC_ENABLE_CODEVELOP`` (optional) one of ``OFF`` (default), ``ON``

>   when ``ON``, indicates that axom and mfem should be built as part of serac's CMake project.

Reviewer: ```suggestion
  When ``ON``, indicates that Axom and MFEM should be built as part of Serac's CMake project.
```

  This can be useful for debugging those libraries, or making changes to them without needing

>   to rebuild all of serac's dependencies through ``spack``.

Reviewer: ```suggestion
  to rebuild all of serac's dependencies through Spack.
```

``ENABLE_ASAN``: (optional) one of ``OFF`` (default), ``ON``
  Enables the Address Sanitizer for memory safety inspections

``-G <generator>``: (optional) typically one of ``Unix Makefiles`` (default), or ``Ninja``

>   which  the Address Sanitizer for memory safety inspections

Reviewer: ```suggestion
  Changes the build system generated by CMake
```

-----------

> the following configure flags are mainly intended for serac developers:

Reviewer: ```suggestion
The following configure flags are mainly intended for Serac developers:
```

``ENABLE_DOCS`` (optional) one of ``OFF`` (default), ``ON``

>   specifying that the documentation targets (sphinx, doxygen) should be built

Reviewer: ```suggestion
  Enables building the documentation
```

>   (requires `sphinx`, `doxygen` which can be built as part of serac's devtools)

Reviewer: ```suggestion
  (requires `sphinx`, `doxygen` which can be built as part of Serac's devtools)
```

> ``ENABLE_CLANGFORMAT`` (optional) one of ``OFF`` (default), ``ON``
>   adds a target named "style" that applies `clang-format` to the sources in
>   serac, to ensure that they conform to the style guide
>   (requires `clang-format` v10.0, which can be built as part of serac's devtools)

Reviewer: ```suggestion
``ENABLE_CLANGFORMAT`` (optional) one of ``OFF`` (default), ``ON``
  Enables the build target named ``style`` and ``check`` that applies ``clang-format`` to the sources in
  Serac, to ensure that they conform to the style guide
  (requires ``clang-format`` v10.0, which can be built as part of Serac's devtools)
```

``CLANGFORMAT_EXECUTABLE`` (optional)

>   used to manually specify which `clang-format` executable to use when configuring with 

Reviewer: ```suggestion
  Used to manually specify which ``clang-format`` executable to use when configuring with 
```

  ``-DENABLE_CLANGFORMAT=ON``

``SPHINX_EXECUTABLE`` (optional)

>   used to manually specify which `sphinx` executable to use when configuring with ``-DENABLE_DOCS=ON``

Reviewer: ```suggestion
  Used to manually specify which ``sphinx-build`` executable to use when configuring with ``-DENABLE_DOCS=ON``
```

``DOXYGEN_EXECUTABLE`` (optional)

>   used to manually specify which `doxygen` executable to use when configuring with ``-DENABLE_DOCS=ON``

Reviewer: ```suggestion
  Used to manually specify which ``doxygen`` executable to use when configuring with ``-DENABLE_DOCS=ON``
```

------

> We also provide a python script ``config-build.py`` that will create a build
> directory and invoke CMake for you. Some example use cases are:

Reviewer: ```suggestion
We also provide a python script ``config-build.py`` that simplify configuring our CMake build.  It does the following:

* Creates a new build and install directory for you. Deleting them if they are present.
* Invoke CMake for you.
* Default to a host-config on known machines.
* Use a specific, tested CMake executable defined in the host-config.

Some example use cases are:
```

Building Serac
--------------

> Serac is a CMake project. Configure with at least the ``-C``

Reviewer: One of the large benefits of `config-build.py` is that it takes the CMake executable from the host-config. Unfortunately, CMake has wildly differing behavior based on what CMake executable is used.

This section needs to say to refer to the host-config on the given platform to see what CMake to use.

Author: > CMake has wildly differing behavior based on what CMake executable is used.

Would it be beneficial to require a more recent version of CMake, so that there are fewer versions to worry about? 3.18 or later, would also have the benefit of native CUDA language support. I haven't noticed any issues with serac + cmake v3.20 - 3.26.

Reviewer: Unfortunately newer CMake's are not always better. It is often the case that their behavior changes and bugs are reintroduced.

The only safe thing I have found is to lock it down to a blessed version of CMake to avoid wasting time at the developer level.

Author: I'll include that in the quickstart guide, then. What is the blessed version of CMake?

Reviewer: on LC its 3.18.0

flag below, to ensure CMake can locate the dependencies installed by spack:

.. code-block:: bash

>   $ cmake . -B build \

Reviewer: In the past, having the source directory, '.' in this case, before the initial cache, '-C', caused a lot of very bad behavior in CMake. I think what happened is it delays loading the initial cache. I can't seem to find any articles to explain this. I am not sure this is still the case, but I would not trust CMake to do the correct thing. 

Please, change this line to have the `.` at the very end of the command line.

            -C path/to/serac/<machine>-<type>-<compiler>.cmake \
            -DCMAKE_BUILD_TYPE=Release

LLNL developers working on one of the LC machines can instead use one of 
the provided CMake cache files in serac's ``host-configs`` directory:

``-C path/to/serac/host-configs/<machine>-<type>-<compiler>.cmake``

----------
 
Some other frequently-useful configure flags

> ``-B <build-directory>`` (recommended) 

Reviewer: Consider using a Directive List

https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#definition-lists

Reviewer: Here is what it looks like:

https://llnl-blt.readthedocs.io/en/develop/api/target.html#blt-add-executable

I use it in BLT to document parameters.

Author: I think that's what I'm using in there right now, just that the entries are in backticks to make it use a monospaced font

> ``-B <build-directory>`` (recommended) 

Reviewer: This is not just recommended, it is required. We do not allow in-source builds, which leaving off the `-B` option turns into.

Author: Sure. I meant that the common pattern of
`mkdir build && cd build && cmake .. <args>`
doesn't require the -B flag, but I'll clarify.

   $ python3 ./config-build.py -hc host-configs/<machine name>-<SYS_TYPE>-<compiler>.cmake

   # If you'd like to configure specific build options, e.g., a release build

>    $ python3 ./config-build.py -hc /path/to/host-config.cmake -bt Release <more CMake build options...>

Reviewer 2: Do we want to keep the note that they should contact me if people want to use the preinstalled TPLs on LC @white238 ?

Note:

>   - ``config-build.py`` will delete the contents of build/install directories if they exist already

Reviewer 2: 👍

> Note:
>   - ``config-build.py`` will delete the contents of build/install directories if they exist already
>   - some flags parsed by ``config-build.py`` use different names than their CMake counterparts,
>     try running ``python3 config-build.py --help`` for more information.

Reviewer: ```suggestion
.. note::

  ``config-build.py`` will delete the contents of build/install directories if they exist already.
  Some flags parsed by ``config-build.py`` use different names than their CMake counterparts,
  try running ``python3 config-build.py --help`` for more information.
```
